File: haskell/grains/src/Grains.hs
module Grains (square, total) where

square :: Integer -> Maybe Integer
square n
    | n > 0 && n < 65 = Just ((generateSquares [])!!index)
    | otherwise = Nothing
    where index = (fromIntegral n - 1)

total :: Integer
total = sum $ generateSquares []

generateSquares :: [Integer] -> [Integer]
generateSquares [] = generateSquares [1]
generateSquares squares@(x:_)
    | length squares == 64 = reverse squares
    | otherwise = generateSquares ([(x * 2)] ++ squares)

File: python/grains/grains.py
squares = [ 2 ** x for x in range(64) ]

def on_square(square):
    _raise_if_invalid(square)
    return squares[square - 1]

def total_after(square):
    _raise_if_invalid(square)
    return sum(squares[:square])

def _raise_if_invalid(square):
    if not square in range(1, 65):
	    raise ValueError('Square out of range')
